agent/src/com/thoughtworks/go/agent/AgentWebSocketClientController.java

```diff
@@ -0,0 +1,272 @@
+import org.eclipse.jetty.websocket.api.extensions.Frame;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.InputStream;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+@WebSocket
+public class AgentWebSocketClientController extends AgentController {
+    private static final Logger LOG = LoggerFactory.getLogger(AgentWebSocketClientController.class);
+    private final SslInfrastructureService sslInfrastructureService;
+    private final GoArtifactsManipulator manipulator;
+    private HttpService httpService;
+    private WebSocketClientHandler webSocketClientHandler;
+    private WebSocketSessionHandler webSocketSessionHandler;
+    private AtomicReference<BuildSession> buildSession = new AtomicReference<>();
+    private JobRunner runner;
+    private PackageAsRepositoryExtension packageAsRepositoryExtension;
+    private SCMExtension scmExtension;
+    private TaskExtension taskExtension;
+    private BuildRepositoryRemote server;
+
+    public AgentWebSocketClientController(BuildRepositoryRemote server, GoArtifactsManipulator manipulator,
+                                          SslInfrastructureService sslInfrastructureService, AgentRegistry agentRegistry,
+                                          AgentUpgradeService agentUpgradeService, SubprocessLogger subprocessLogger,
+                                          SystemEnvironment systemEnvironment, PluginManager pluginManager,
+                                          PackageAsRepositoryExtension packageAsRepositoryExtension, SCMExtension scmExtension,
+                                          TaskExtension taskExtension, HttpService httpService,
+                                          WebSocketClientHandler webSocketClientHandler, WebSocketSessionHandler webSocketSessionHandler) {
+        super(sslInfrastructureService, systemEnvironment, agentRegistry, pluginManager, subprocessLogger, agentUpgradeService);
+        this.server = server;
+        this.manipulator = manipulator;
+        this.packageAsRepositoryExtension = packageAsRepositoryExtension;
+        this.scmExtension = scmExtension;
+        this.taskExtension = taskExtension;
+        this.sslInfrastructureService = sslInfrastructureService;
+        this.httpService = httpService;
+        this.webSocketClientHandler = webSocketClientHandler;
+        this.webSocketSessionHandler = webSocketSessionHandler;
+    }
+
+    @Override
+    public void ping() {
+        // Do nothing
+    }
+
+    @Override
+    public void execute() {
+        // Do nothing
+    }
+
+    @Override
+    public void work() throws Exception {
+        if (sslInfrastructureService.isRegistered()) {
+            if (webSocketSessionHandler.isNotRunning()) {
+                webSocketSessionHandler.clearCallBacks();
+                synchronized (this) {
```

Comment:
Why is the sync required? If this is done to synchronize on session, then shouldn't all the three calls to `webSocketSessionHandler` be in the same synchronized block, and may be we should be double checking on `webSocketSessionHandler.isNotRunning()` as well.